### PR TITLE
feat: support gradient clipping in PyTorchTrial via callbacks [DET-3233]

### PR DESCRIPTION
**Pull request by aaron276h**
## Description
Previously we supported gradient clipping in PyTorch only via special hyperparameters. Moving to supporting it via callbacks will make it more transparent for users as well as allow users to apply more specialized clipping techniques.

As part of this PR also providing pre-made callbacks for the two types of gradient clipping we had previously supported. 

[DET-3233](https://determinedai.atlassian.net/browse/DET-3233)

## Test Plan
Updated existing tests.

### CODE REVIEW

File: harness/tests/experiment/fixtures/pytorch_xor_model.py [hunk 1]
```diff
@@ -51,22 +51,22 @@ def __init__(self, context):
             nn.Linear(context.get_hparam("hidden_size"), 1),
             nn.Sigmoid(),
         )
-        reset_parameters(self.main_net)
+        pytorch.reset_parameters(self.main_net)
```
Comment (shiyuann) [inline harness/tests/experiment/fixtures/pytorch_xor_model.py:54]: Question: A very bad consequence of importing functions using only one path and using module would be a situation like this. It looks like we are importing from Pytorch. Maybe it's a place we should use `det_torch` or other options. Or maybe we should just import functions directly without using a module name. @rb-determined-ai Thoughts on it?

File: harness/determined/pytorch/_pytorch_trial.py [hunk 2]
```diff
@@ -416,7 +398,15 @@ def _train_for_step(self, step_id: int, batches_per_step: int) -> workload.Respo
                         parameters=parameters, divisor=self.hvd_config.aggregation_frequency
                     )
 
-                self._clip_grads(parameters)
+                # TODO: Remove this check in v0.12.8.
```
Comment (shiyuann) [inline harness/determined/pytorch/_pytorch_trial.py:401]: blocking: This looks like a breaking change since we no longer use these hparams. Should go through the breaking change process as in https://docs.google.com/document/d/1P4ym1eF3LosYe_hGAtFqXyt8PhyPqTmEihiPtG8tnm8/edit.

Comment (aaron276h) [reply harness/determined/pytorch/_pytorch_trial.py:401]: Updated.

File: harness/determined/pytorch/_callback.py [hunk 3]
```diff
@@ -39,6 +41,14 @@ def on_train_step_end(self, step_id: int, metrics: Dict[str, Any]) -> None:
         """
         pass
 
+    def on_before_optimizer_step(self, parameters: Generator) -> None:
```
Comment (rb-determined-ai) [inline harness/determined/pytorch/_callback.py:44]: Can we rename this `on_optimizer_step_start()` to be consistent with e.g. `on_validation_step_start()` or `before_optimizer_step()` to be simple?

`on_before_optimizer_step()` sounds super awkward.

File: harness/tests/experiment/fixtures/pytorch_xor_model.py (hunk 1 above)
Comment (rb-determined-ai) [reply harness/tests/experiment/fixtures/pytorch_xor_model.py:54]: I'm ok with this, but I would also be ok with:
```
import determined as det
import determined.pytorch
det.pytorch.reset_parameters()
```

The reason I'm ok with what we have now is that it is consistent with the google python coding standards, and it's awfully nice to stick to a published standard and not fight over these things.  If I was going to break the standard though, I would just not import the submodule at all.  Using the full path module path works fine and isn't any longer.

File: harness/determined/pytorch/_callback.py [hunk 4]
```diff
@@ -75,3 +85,29 @@ def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
...
+class ClipGradsL2Norm(PyTorchCallback):
+    """
+    Callback that performs gradient clipping using
+    `L2 Norm <https://pytorch.org/docs/stable/nn.html#clip-grad-norm>`_.
+    """
+
+    def __init__(self, clip_value: float) -> None:
+        self._clip_value = clip_value
+
+    def on_before_optimizer_step(self, parameters: Generator) -> None:
```
Comment (rb-determined-ai) [inline harness/determined/pytorch/_callback.py:99]: Let's make this an Iterator, not a Generator, since a Generator is more specific, and also not actually accurate here.

Also in other places.

File: harness/determined/pytorch/_callback.py (hunk 3 above)
Comment (rb-determined-ai) [reply harness/determined/pytorch/_callback.py:44]: IMO, naming patterns are helpful in bringing clarity, until they make things less clear.  And I think this is a place where following the pattern brings less clarity.

I won't block this change over this though.

File: harness/determined/pytorch/_callback.py (hunk 4 above)
Comment (aaron276h) [reply harness/determined/pytorch/_callback.py:99]: Their docs claim to return an iterator, but inspecting their source code they actually return  a Generator from `model.parameters()`. But happy to update this.
